Pinch slider orientation

artsouflMS opened the pull request:
## Overview

Fix bug #6756 PinchSlider orientation doesn't change when setting Slider Axis to Y or Z

Updated PinchSlider.cs to update the orientation of TrackVisuals, TickMarks and ThumbRoot depending on the sliderAxis orientation

## Changes
- Fixes: #6756

OnValidate the property of CurrentSliderAxis is being set and when that happens, the orientation of the visual components is being updated if the axis has been changed.
This is for editor changes
If user change the property of CurrentSliderAxis at runtime, the update function will be called on setter so the visual's orientation will be updated

The UpdateVisualsOrientation() function update the local rotation of the gameobjects
It doesn't change the local position
So if user wants to update the position, then user needs to do this manually in editor or by code at runtime.

I could reset the local position of the gameobjects, so they don't preserve any wrong translation from previous slideraxis orientation. The problem with this approach is that maybe the user changes the axis by mistake and then when user rolls back to previous axis setup then the local translation will be lost and user will need to set it up again.
So, I decided for this reason, to leave the position untouched.

The private variable that the property about PreviousSliderAxis refers to, is nullable for initialization purposes. if the variable hasn't any value then it needs to become equal to the CurrentSliderAxis

I also added 2 more tests in PinchSliderTests
One test is setting the visual components to null and tests that the slider move still works
The other test, checks the orientation of the components while the slideraxis changes.

I haven't updated the documentation, will do that after this behaviour/PR is accepted. 


## Verification
> This optional section is a place where you can detail the specific type of verification 
> you want from reviewers. For example, if you want reviewers to checkout the PR locally
> and validate the functionality of specific scenarios, provide instructions
> on the specific scenarios and what you want verified.
>
> If there are specific areas of concern or question feel free to highlight them here so
> that reviewers can watch out for those issues.
>

The thumb root isn't being rotated while the axis changes
Shouldn't the thumbroot rotate as well?
If we want to update the thumb root rotation, then the HandInteractionExample breaks, as the PinchSlider there has an additional child under the ThumbRoot so that child has wrong orientation if its parent ThumbRoot is rotated to match the SliderAxis
So, do we leave the ThumbRoot as it is, and if the user wants, then user can make changes manually?
It looks visually wrong though for the ThumbRoot to remain in the same orientation.

> As a reviewer, it is possible to check out this change locally by using the following
> commands (substituting {PR_ID} with the ID of this pull request):
>
> git fetch origin pull/{PR_ID}/head:name_of_local_branch
>
> git checkout name_of_local_branch

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/microsoft/MixedRealityToolkit-Unity?pullRequest=6858) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/microsoft/MixedRealityToolkit-Unity?pullRequest=6858) before we can accept your contribution.<br/><br/>:x: artsouflMS [sign now](https://cla.opensource.microsoft.com/microsoft/MixedRealityToolkit-Unity?pullRequest=6858)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/microsoft/MixedRealityToolkit-Unity?pullRequest=6858) it.</sub>

keveleigh commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Sliders/PinchSlider.cs:
```diff
@@ -53,8 +85,35 @@ public float SliderValue
+
         [Serializable]
-        private enum SliderAxis
+        public enum SliderAxis
```
If possible, it'd be great to move this newly `public` enum [into its own file](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/CodingGuidelines.html#public-classes-structs-and-enums-should-all-go-in-their-own-files).

artsouflMS commented:
thank you for the feedback
I will move the SliderAxis enum into a new file

@julenka 
about the second case:

this is the slider in x axis 
![Capture1](https://user-images.githubusercontent.com/42405657/70705892-d0310100-1ccc-11ea-9fad-139ea3ef48d2.JPG)

this is what happens when changing it into Y axis 
![Capture2](https://user-images.githubusercontent.com/42405657/70706031-09697100-1ccd-11ea-9ae7-170e9ec587ca.JPG)

is this what you would expect when changing it into Y axis?
this has modified the Cellheight into what the CellWidth was and has the local position rest to zero
![Capture3](https://user-images.githubusercontent.com/42405657/70706105-33229800-1ccd-11ea-8095-b092b8389d3d.JPG)

if this last one isn't what you would expect.
Could you describe what would you expect?

julenka commented:
Artemis, the below is what i would expect, so yes "the last one" is what I expect
![image](https://user-images.githubusercontent.com/168492/70771717-eff51300-1d26-11ea-9595-e69ab4012bd5.png)

> Maybe we could rotate the child Button_AnimationContainer? So the rotation applies only to that gameobject?

I was thinking you could add a third field "thumb visual" that would point just to the thumb visual, not to the parent that also contains the text. It would nice to put all of these under "slider visuals" or "slider axis visuals" header, then I would see something like this (in editor form)

### Slider axis visuals
* Track Visual
* Tick Visual
* Thumb Visual (with a tooltip that says "point this to just the visual component of the thumb -- this will get rotated to match the slider axis")

artsouflMS commented:
Hi @julenka 

Thank you for your help and feedback,

I committed the changes

- CellWidth and CellHeight are updated then changing slider axis in order to match what was previously set to.
-Local position of gameobjects is being reset as well on slider axis change
-Added a new property ThumbVisuals which is only the thumb gameobject. This is rotated to match the slider axis rotation. A difference with the track visuals and tick marks, is that thumb visuals global rotation is set and not local rotation as with the other gameobjects. 
The reason for this is that, if we set the localrotation then in scenes like the PinchExample that have a local rotation on a parent of the thumb, then these scenes will be broken. 
To me it makes more sense to update the local rotation. So user could add manually a rotation in editor that might be desirable for the whole pinch slider.
But if we could with this solution then we will need to update all scenes that have pinch slider.
What do you think?

Kind regards
Artemis

keveleigh commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Sliders/PinchSlider.cs:
```diff
@@ -31,7 +31,7 @@ public GameObject ThumbRoot
                 InitializeSliderThumb();
             }
-        }
+        } 
```
```suggestion
        }
```

artsouflMS replied:
Not sure what's the recommended change is?

keveleigh replied:
Ah just removing an extra space at the end of the line. Not super important!

artsouflMS replied:
ok, I made this change as well :)

keveleigh commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Sliders/PinchSlider.cs:
```diff
@@ -48,17 +48,87 @@ public float SliderValue
+        }
+
+        private SliderAxis? previousSliderAxis = null;
+        public SliderAxis PreviousSliderAxis
```
Does this need to be `public`? It seems like local state management that we wouldn't necessarily want other scripts arbitrarily setting.

artsouflMS replied:
I ve changed it to private.

keveleigh commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Sliders/PinchSlider.cs:
```diff
@@ -48,17 +48,87 @@ public float SliderValue
+        [Tooltip("The gameObject that contains the thumb visual. This will get rotated to match the slider axis")]
+        [SerializeField]
+        private GameObject thumbVisual = null;
+        public GameObject ThumbVisual
```
Please add summary tags for these new public properties. They can match the above Tooltip text

artsouflMS replied:
Done

julenka commented:
> But if we could with this solution then we will need to update all scenes that have pinch slider.

Local rotation also makes more sense to me. It's not a problem to change scenes. However, to keep the change scoped maybe we should break that into a different change (so this doesn't balloon to changing too much). I would be fine if we backed out the change to add the thumb and just created a follow up issue referencing this PR.

julenka commented:
One more comment -- if you could please add documentation to [README_Sliders.md](https://github.com/microsoft/MixedRealityToolkit-Unity/blob/mrtk_development/Documentation/README_Sliders.md) describing the new behavior that would be great. This is part of our guidelines [pull requests should contain documentation for new feature](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/Contributing/PullRequests.html#documentation-should-be-added-in-the-same-pull-request-as-a-fix--feature). Specifically it would be good to mention when axis is updated, visuals are updated if they are specified. Please also add explanation of new fields.

artsouflMS commented:
Hi @julenka 

I updated the documentation by adding at the end explanation of inspector properties and after that I added explanation of when the track visuals and tick marks are updated

I removed the thumb visuals property and functionality related to updating its transform

Also made the suggested changes from @keveleigh 

Thank you both for the feedback
Artemis

keveleigh commented on Assets/MixedRealityToolkit.SDK/Features/UX/Scripts/Sliders/SliderAxis.cs:
```diff
@@ -0,0 +1,18 @@
+    /// Describes in which axis to orient the slider
+    /// </summary>
+    [Serializable]
+    public enum SliderAxis
```
Ah, along the lines of https://github.com/microsoft/MixedRealityToolkit-Unity/pull/6924#discussion_r358936001, we should definitely try to standardize out axis enums. Don't think it's worth blocking this PR right now though

keveleigh commented:
/azp run
